File: Exercism/python/yacht/yacht.py
from collections import Counter
from functools import partial


def _n_of_a_kind(n, op, dice):
    which, how_many = Counter(dice).most_common(1)[0]
    if how_many >= n:
        return op(which)
    return 0


YACHT = partial(_n_of_a_kind, 5, lambda _: 50)


def _score_n(n, dice):
    return n * Counter(dice).get(n, 0)


ONES = partial(_score_n, 1)
TWOS = partial(_score_n, 2)
THREES = partial(_score_n, 3)
FOURS = partial(_score_n, 4)
FIVES = partial(_score_n, 5)
SIXES = partial(_score_n, 6)


def FULL_HOUSE(dice):
    if sorted(Counter(dice).values()) == [2, 3]:
        return sum(dice)
    return 0


FOUR_OF_A_KIND = partial(_n_of_a_kind, 4, lambda which: 4 * which)


def _straight(start, dice):
    if sorted(dice) == list(range(start, start+5)):
        return 30
    return 0


LITTLE_STRAIGHT = partial(_straight, 1)
BIG_STRAIGHT = partial(_straight, 2)


def CHOICE(dice):
    return sum(dice)


def score(dice, category):
    return category(dice)

File: Exercism/haskell/yacht/src/Yacht.hs
module Yacht (yacht, Category(..)) where

import Data.List (sort)

import Counter (count, allCountsByFrequency)

data Category = Ones
              | Twos
              | Threes
              | Fours
              | Fives
              | Sixes
              | FullHouse
              | FourOfAKind
              | LittleStraight
              | BigStraight
              | Choice
              | Yacht
              deriving (Eq, Enum, Show)

type DiceFace = Int
type Dice = [DiceFace]
type Score = Int

yacht :: Category -> Dice -> Score

yacht Yacht dice = 10 * computeNofAKind 5 dice

yacht FullHouse dice = if isFullHouse then sum dice else 0
  where isFullHouse = map snd (allCountsByFrequency dice) == [2, 3]

yacht FourOfAKind dice = 4 * computeNofAKind 4 dice

yacht LittleStraight dice = streightScoreFrom 1 dice

yacht BigStraight dice = streightScoreFrom 2 dice

yacht Choice dice = sum dice

yacht category dice
  | category `elem` [(Ones)..Sixes] = computeOnesToSixes category dice
  | otherwise = 0

computeOnesToSixes :: Category -> Dice -> Score
computeOnesToSixes category dice = (*) <$> id <*> flip count dice $ fromEnum category + 1

computeNofAKind :: Int -> Dice -> Score
computeNofAKind n dice = score_for_n $ last . allCountsByFrequency $ dice
  where score_for_n (which, how_many)
          | how_many >= n = which
          | otherwise     = 0

streightScoreFrom :: Int -> Dice -> Score
streightScoreFrom n dice
  | sort dice == [n .. n + 4] = 30
  | otherwise = 0
